seqr/views/utils/permissions_utils.py

```diff
@@ -53,18 +72,38 @@ def check_multi_project_permissions(obj, user):
     raise PermissionDenied("{user} does not have view permissions for {object}".format(user=user, object=obj))
 
 
+def _get_workspaces_user_can_view(user):
+    # the 'service_account_session' below will be replaced by 'session' after seqr client ID is whitelisted
+    is_staff_user = user.is_staff
+    session = service_account_session
+    requested_fields = 'public,workspace.name,workspace.namespace,workspace.workspaceId'
+    workspace_list = session.list_workspaces(requested_fields)
+    workspaces = []
+    for ws in workspace_list:
+        if not ws['public']:
+            if is_staff_user:
+                workspaces.append('/'.join([ws['workspace']['namespace'], ws['workspace']['name']]))
+            else:
+                acl = session.get_workspace_acl(ws['workspace']['namespace'], ws['workspace']['name'])
+                if user.email in acl.keys():
+                    workspaces.append('/'.join([ws['workspace']['namespace'], ws['workspace']['name']]))
+    return workspaces
+
+
 def get_projects_user_can_view(user):
-    can_view_filter = Q(can_view_group__user=user)
+    if anvilSessionStore.get_session(user):
+        workspaces = _get_workspaces_user_can_view(user)
+        can_view_filter = (Q(can_view_group__user=user) & Q(workspace__isnull=True)) | Q(workspace__in=workspaces)
```

Comment:
Okay, I figured out how you can do the query you want if  workspace name and namespace are stored as separate fields!

```
from django.db.models.functions import Concat
from django.db.models import Value
...
can_view_filter = Q(can_view_group__user=user)
if user.is_staff:
   can_view_filter = can_view_filter | Q(disable_staff_access=False)

if anvilSessionStore.get_session(user):
   workspaces = _get_workspaces_user_can_view(user)
   anvil_projects = Project.objects.annotate(workspace=Concat('workspace_namespace', Value('/'), 'workspace_name')).filter(workspace__in=workspaces)
  local_projects = Project.objects.filter(can_view_filter & Q(workspace__isnull=True)
   return (anvil_projects | local_projects).distinct()
else:
   return Project.objects.filter(can_view_filter)

Reply:
Got it. Thanks.